File: android/app/src/main/java/com/cloudpad/controller/Protocol.kt
package com.cloudpad.controller
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.roundToInt

object Protocol { const val VERSION:Byte=1; const val PORT=26760; const val PACKET_SIZE=61; const val TOKEN_SIZE=32 }
object Buttons { const val A=1 shl 0;const val B=1 shl 1;const val X=1 shl 2;const val Y=1 shl 3;const val LB=1 shl 4;const val RB=1 shl 5;const val BACK=1 shl 6;const val START=1 shl 7;const val L3=1 shl 8;const val R3=1 shl 9;const val UP=1 shl 10;const val DOWN=1 shl 11;const val LEFT=1 shl 12;const val RIGHT=1 shl 13 }
data class PadState(val lx:Float=0f,val ly:Float=0f,val rx:Float=0f,val ry:Float=0f,val lt:Float=0f,val rt:Float=0f,val buttons:Int=0)
fun axisShort(v:Float):Short { val c=v.coerceIn(-1f,1f);return if(c<=-1f)Short.MIN_VALUE else (c*Short.MAX_VALUE).roundToInt().toShort() }
fun deadzone(v:Float,d:Float):Float { val a=abs(v);return if(a<=d)0f else kotlin.math.sign(v)*(a-d)/(1f-d) }
fun packet(session:Int,sequence:Int,time:Long,state:PadState,token:ByteArray):ByteArray { require(token.size==32);return ByteBuffer.allocate(Protocol.PACKET_SIZE).order(ByteOrder.LITTLE_ENDIAN).apply { put(Protocol.VERSION);putInt(session);putInt(sequence);putLong(time);putShort(axisShort(state.lx));putShort(axisShort(state.ly));putShort(axisShort(state.rx));putShort(axisShort(state.ry));put((state.lt.coerceIn(0f,1f)*255).roundToInt().toByte());put((state.rt.coerceIn(0f,1f)*255).roundToInt().toByte());putShort(state.buttons.toShort());put(token) }.array() }

File: windows-receiver/CloudPad.Protocol/Protocol.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace CloudPad.Protocol;

public static class ProtocolConstants
{
    public const byte Version = 1;
    public const int DefaultPort = 26760;
    public const int PacketSize = 61;
    public const int TokenSize = 32;
    public const int DefaultTimeoutMs = 500;
}

[Flags]
public enum GamepadButtons : ushort
{
    None=0, A=1<<0, B=1<<1, X=1<<2, Y=1<<3, LeftShoulder=1<<4,
    RightShoulder=1<<5, Back=1<<6, Start=1<<7, LeftThumb=1<<8,
    RightThumb=1<<9, DPadUp=1<<10, DPadDown=1<<11, DPadLeft=1<<12, DPadRight=1<<13
}

public readonly record struct ControllerState(float LeftX, float LeftY, float RightX, float RightY,
    float LeftTrigger, float RightTrigger, GamepadButtons Buttons)
{
    public static ControllerState Neutral => default;
    public ControllerState Clamp() => new(
        Math.Clamp(LeftX,-1,1), Math.Clamp(LeftY,-1,1), Math.Clamp(RightX,-1,1), Math.Clamp(RightY,-1,1),
        Math.Clamp(LeftTrigger,0,1), Math.Clamp(RightTrigger,0,1), Buttons);
}

public readonly record struct InputPacket(uint SessionId, uint Sequence, long TimestampMs,
    ControllerState State, byte[] Token)
{
    public byte[] Serialize()
    {
        if (Token.Length != ProtocolConstants.TokenSize) throw new ArgumentException("Token must be 32 bytes");
        var b = new byte[ProtocolConstants.PacketSize];
        b[0] = ProtocolConstants.Version;
        BinaryPrimitives.WriteUInt32LittleEndian(b.AsSpan(1), SessionId);
        BinaryPrimitives.WriteUInt32LittleEndian(b.AsSpan(5), Sequence);
        BinaryPrimitives.WriteInt64LittleEndian(b.AsSpan(9), TimestampMs);
        var s=State.Clamp(); int o=17;
        foreach(var v in new[]{s.LeftX,s.LeftY,s.RightX,s.RightY}) { BinaryPrimitives.WriteInt16LittleEndian(b.AsSpan(o), Axis.ToInt16(v)); o+=2; }
        b[o++]=Axis.ToByte(s.LeftTrigger); b[o++]=Axis.ToByte(s.RightTrigger);
        BinaryPrimitives.WriteUInt16LittleEndian(b.AsSpan(o),(ushort)s.Buttons); o+=2;
        Token.CopyTo(b,o);
        return b;
    }

    public static bool TryParse(ReadOnlySpan<byte> b, out InputPacket packet)
    {
        packet=default;
        if(b.Length!=ProtocolConstants.PacketSize || b[0]!=ProtocolConstants.Version) return false;
        int o=17;
        float lx=Axis.FromInt16(BinaryPrimitives.ReadInt16LittleEndian(b[o..])); o+=2;
        float ly=Axis.FromInt16(BinaryPrimitives.ReadInt16LittleEndian(b[o..])); o+=2;
        float rx=Axis.FromInt16(BinaryPrimitives.ReadInt16LittleEndian(b[o..])); o+=2;
        float ry=Axis.FromInt16(BinaryPrimitives.ReadInt16LittleEndian(b[o..])); o+=2;
        float lt=b[o++]/255f, rt=b[o++]/255f;
        var buttons=(GamepadButtons)BinaryPrimitives.ReadUInt16LittleEndian(b[o..]); o+=2;
        packet=new(BinaryPrimitives.ReadUInt32LittleEndian(b[1..]),BinaryPrimitives.ReadUInt32LittleEndian(b[5..]),
            BinaryPrimitives.ReadInt64LittleEndian(b[9..]),new(lx,ly,rx,ry,lt,rt,buttons),b.Slice(o,32).ToArray());
        return true;
    }
}

public static class Axis
{
    public static short ToInt16(float value) { value=Math.Clamp(value,-1,1); return value<=-1?short.MinValue:(short)Math.Round(value*short.MaxValue); }
    public static float FromInt16(short value) => value<0 ? value/32768f : value/32767f;
    public static byte ToByte(float value)=>(byte)Math.Round(Math.Clamp(value,0,1)*255);
    public static float Deadzone(float value,float deadzone)
    { var a=Math.Abs(value); if(a<=deadzone)return 0; return Math.Sign(value)*(a-deadzone)/(1-deadzone); }
}

public static class Sequence
{
    public static bool IsNewer(uint candidate,uint previous)=>candidate!=previous && unchecked(candidate-previous)<0x80000000u;
}

public sealed record Session(uint Id, byte[] Token, string DeviceName);
public sealed class SessionManager
{
    readonly object sync=new(); Session? current;
    public Session Create(string device) { lock(sync) return current=new((uint)RandomNumberGenerator.GetInt32(1,int.MaxValue),RandomNumberGenerator.GetBytes(32),device); }
    public bool Validate(uint id,ReadOnlySpan<byte> token){lock(sync)return current is not null && current.Id==id && CryptographicOperations.FixedTimeEquals(current.Token,token);}
    public bool Clear(uint id){lock(sync){if(current?.Id!=id)return false;current=null;return true;}}
    public void Clear(){lock(sync)current=null;}
}
